Dockerfile

```diff
@@ -1,29 +1,115 @@
-FROM ruby:2.7.2
-
-# 1. Configuring apt to properly install yarn
-# 2. Installing node, postgres-client, yarn
-# 3. Clearing apt-cached files to tighten the image size
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1 apt-key add - \
-  && echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list \
-  && apt-get update -qq \
-  && apt-get install -y build-essential \
-  nodejs \
-  yarn \
-  && rm -rf /var/lib/apt/lists/*
+# Generating a generic Rails image.
+# Just for caching.
+FROM ruby:2.7-alpine AS rails
 
-WORKDIR /app
+# `apk` is like `apt`, but for Alpine Linux.
+RUN apk add --update --no-cache --quiet \
+    build-base \
+    nodejs-current \
+    postgresql-dev \
+    sqlite-dev \
+    tzdata \
+    yarn
+
+# Create a generic Gemfile with only Rails.
+RUN echo -e "source 'https://rubygems.org'\ngem 'rails'" > /Gemfile
+# Install Rails, quietly, in 4 threads.
+RUN bundle install --quiet -j4
+
+# Uses the image above as the base for a builder image.
+# This will **exclude** development and testing gems.
+FROM rails as rails-builder
```

Comment:
I think `rails-builder` should include `yarn install` because `rails s` won't run without yarn installed.

Reply:
And we need the app copied in!

Reply:
The reason I do not run `yarn install` in this step is that `node_modules` exist in `/app/`.  So when `docker-compose.yml` mounts the volume, the existing `/app/node_modules/` is overwritten by the one from the host machine.

We could add the `yarn install --check-files` to the command in `docker-compose.yml` to get a single-command build and run.  This will have an impact on how quickly the server starts, of course.

Reply:
YES!!!! Single-command build and run!